oeps/oep-0023-theming.rst

```diff
@@ -0,0 +1,204 @@
+      absolute minimum. If breaking changes must occur, the core
+      contributor team will enumerate them with every named release of
+      the Open edX platform.
+
+4. The core contributor team will provide official documentation,
+      examples, and tools to help other core contributors and open
+      source community members develop themes.
+
+5. Further modification to UI elements or content will not be officially
+      supported within the scope of this theming system. A system to
+      enable content modification may eventually be developed and
+      supported, but its definition is outside the scope of this
+      proposal.
+
+Plan
+----
+
+1. API definition for the new theming system will be proposed, agreed
+      upon, and documented.
+
+2. The core contributor team will develop an example theme compliant
+      with the new system and use it to build support into the platform.
+      Existing work [8] on such a theme already exists in the
+      edx-bootstrap repository.
+
+3. Open edX core and community contributors will build scaffolding to
+      support the new theming system throughout the Open edX platform.
+      This scaffolding will be implemented in as unobtrusive a means as
+      possible with all attempts made not to interfere with existing
+      theming systems.
+
+4. Once scaffolding is complete, the core contributor team will announce
+      an alpha release of the new theming system with the closest named
+      release of the Open edX platform.
+
+5. Throughout the lifecycle of the aforementioned release, the core
+      contributor team will solicit feedback and contributions and
+      respond to suggestions from community members willing to try out
+      the new theming system.
+
+6. By the next sequential named release of the Open edX platform,
+      assuming the new theming system has delivered value to its alpha
+      stakeholders, the core contributor team will announce a beta
+      release of the new theming system and announce the official
+      deprecation of all legacy theming systems.
+
+7. The Open edX core contributor team will continue to solicit
+      contributions and feedback on the theming system until the
+      following named release of the platform, by which point they will
+      announce official support for the new system.
+
+8. Removal of legacy theming-related code may begin two named releases
+      after the announcement of its deprecation.
+
+Consequences
+============
+
+-  All existing theming systems will be deprecated and support will
+      eventually be removed from the platform.
+
+-  Theming scope will be restricted to look-and-feel and branding
+      element modification. Content modification must take place outside
+      the scope of theming.
+
+-  The Open edX platform will gain further ability to evolve its
+      frontend architecture.
+
+-  The Open edX platform’s static asset size and build times will be
+      greatly reduced.
```

Comment:
Great to see this kicking off! I agree we need a single system that would be properly documented and maintained, however using only Bootstrap theming (only visual part) is, in my opinion, not going to be sufficient for theming needs that vendors and maintainers of instances have. There are a lot of situations that require changing the content of pages, with one of the most obvious ones being setting the contents of static pages and changing the header and footer menu items (those are just the lightest modifications that come to mind, there are a lot more).

I don't see Comprehensive Theming as a system that doesn't cover a lot of needs - the template overriding system is something we definitely need and use. The problem lies more in usage of different types of templates (underscore) and lack of a proper API/documentation on how to include certain elements of the platform into custom templates. So doing customisations and implementing our custom templates is not a process where we refer to docs, but we rather figure out how everything works in the default templates and replicate it in our own. So when further work is done on the frontend by the core team and released in the next named release, a lot of stuff breaks because, naturally, the core team doesn't know what we use and what we don't (and I really wouldn't expect they do, after all we're essentially hacking stuff all around in our work).

What would certainly be helpful is a system where templates are clear and can be overridden (in a way we can override them using Comprehensive theming), where it's very clear and documented how to implement the LMS features in custom templates (for example: _You can call the list of 5 newest courses on the platform using X, passing these arguments and expecting this response_), a centralised place where default SASS variables can be overridden in the theme and documented classes that can be reused (I guess Bootstrap would answer this). A prerequisite for this would, I guess, be a cleanup of existing templates in the platform.

My question would be - what is the relation between this and the planned effort of complete separation of frontend and backend? Regarding how closely related to those future plans and the future architecture the new theming engine is, Is it worthwhile to scrap everything and build a new theming engine from scratch, or maybe just improve the Comprehensive theming engine as an intermediate solution before the full transition to a separated frontend?

I have also started a project to explore the possibilities of a future completely separated LMS (including the caveats, and "what do we need to make it work"): https://github.com/appsembler/react-lms
It is in a very early stage, however as you can see I have addressed theming in it in a very similar fashion to current Comprehensive theming - maybe/hopefully it helps in your future explorations for the platform.

Reply:
@grozdanowski excellent points about the utility of comp theming template overrides -- as others have raised this as well, I'm addressing below in a broader comment. As much as I'd love to scrap the legacy systems and concentrate on building the new system instead, internal resourcing constraints will mean this needs to roll out gradually, rather than an all-in-one rebuild.

I envisioned the new system with the primary intention of working with the newer (decoupled frontend) architecture pattern. Achieving compatibility with legacy code may involve some hacky one-off integrations, but since the goal is to eliminate legacy frontends anyway, supporting them is less of a higher-order concern.

Your React LMS has actually been a major rallying point for us internally -- especially considering the timeline within which you completed it! We will definitely check out your theming system as well; looking forward to gathering insight and inspiration from your approach.